Sources/NIOFileSystem/CopyStrategy.swift

```diff
@@ -0,0 +1,129 @@
+    // a much more complex internal state machine could allow doing two of these if desired
+    // This may not result in a faster copy though so things are left simple
+    internal static func determinePlatformDefault() -> Wrapped {
+        #if os(macOS) || os(Linux) || os(Windows)
+        // 4 concurrent file copies/directory scans.
+        // Avoiding storage system contention is the dominant aspect here.
+        // Empirical testing on an SSD copying to the same volume with a dense directory of small
+        // files and sub directories of similar shape totalling 12GB showed improvements in elapsed
+        // time for (expected) increases in CPU time up to parallel(8), beyond this the increases
+        // in CPU came with only moderate gains.
+        // Anyone tuning this is encouraged to cover worst case scenarios.
+        return .parallel(8)
+        #elseif os(iOS) || os(tvOS) || os(watchOS) || os(Android)
+        // Reduced maximum descriptors in embedded world
+        // This is chosen based on biasing to safety, not empirical testing.
+        return .parallel(4)
+        #else
+        // Safety first, if we have no view on it keep it simple.
+        return .sequential
+        #endif
+    }
+}
+
+extension CopyStrategy {
+    // A copy fundamentally can't work without two descriptors unless you copy
+    // everything into memory which is infeasible/inefficeint for large copies.
+    private static let minDescriptorsAllowed = 2
+
+    /// Operate in whatever manner is deemed a reasonable default for the platform.
+    /// This will limit the maximum file descriptors usage based on 'reasonable' defaults.
+    ///
+    /// Current assumptions (which are subject to change):
+    /// - Only one copy operation would be performed at once
+    /// - The copy operation is not intended to be the primary activity on the device
+    public static let platformDefault: Self = Self(Self.determinePlatformDefault())
+
+    /// The copy is done asynchronously, but only one operation will occur at a time.
+    /// This is the only way to guarantee only one callback to the `shouldCopyItem` will happen at a time
+    public static let sequential: Self = Self(.sequential)
+
+    /// The minimum possible parallelism supported by ``parallel(maxDescriptors:)``
+    /// Using this in your non testing code paths would be an anti pattern, use ``sequential``instead.
+    ///
+    /// Less than this and you can't do a copy (without pulling an entire file into memory, which is not generally possible, or desirable)
+    /// This is is not quite the same as sequential, different code paths are used.
+    ///
+    /// This is primarily exposed for testing to ensure use of the parallel paths (which are more complex) while keeping actual
+    /// parallelism to minimal levels to make debugging simpler.
+    public static let minimalParallel: Self = .init(.parallel(minDescriptorsAllowed))
```

Comment:
Is this useful outside of our own tests?

Reply:
I found writing tests using ```.parallel(maxDescriptors:2)``` to be painful since that call declares throws, necessitating lots of cascading the throwing capability or wrapping it in an ugly try/catch.

It seemed reasonable that other people testing code using this might want to make their tests follow the parallel path via ```.platformDefault```, but have the option to switch to the limited parallelism easily without having to mess about at the CopyStrategy declaration point in their tests.

Happy to hide it and just push it into our tests if you prefer.

Reply:
I don't feel too strongly here but my lean is to hide it from the public API, as you note using it outside of testing is an anti-pattern, and I think it's unlikely that users will also need this in their tests. If they do they can `try!` `parallel(maxDescriptors: 2)` (which should be safe per our docs).

Reply:
cool, I'll move it to the testing